Sources/AsyncAlgorithms/Debounce/AsyncDebounceSequence.swift

```diff
@@ -0,0 +1,115 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension AsyncSequence {
+    /// Creates an asynchronous sequence that emits the latest element after a given quiescence period
+    /// has elapsed by using a specified Clock.
+    @available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+    public func debounce<C: Clock>(for interval: C.Instant.Duration, tolerance: C.Instant.Duration? = nil, clock: C) -> AsyncDebounceSequence<Self, C> where Self: Sendable {
+        AsyncDebounceSequence(self, interval: interval, tolerance: tolerance, clock: clock)
+    }
+
+    /// Creates an asynchronous sequence that emits the latest element after a given quiescence period
+    /// has elapsed.
+    @available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+    public func debounce(for interval: Duration, tolerance: Duration? = nil) -> AsyncDebounceSequence<Self, ContinuousClock> where Self: Sendable {
+        self.debounce(for: interval, tolerance: tolerance, clock: .continuous)
+    }
+}
+
+/// An `AsyncSequence` that emits the latest element after a given quiescence period
+/// has elapsed.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct AsyncDebounceSequence<Base: AsyncSequence, C: Clock>: Sendable where Base: Sendable {
+    /// This class is needed to hook the deinit to observe once all references to the ``AsyncDebounceSequence`` are dropped.
```

Comment:
why is the lifespan of the AsyncSequence (the plan on HOW to do things) going to terminate the iteration? im not sure I follow on why this is appropriate behavior.

Reply:
You are right that doesn't make sense here. This is some of the behaviour I inherited from the implementation of Source `AsyncSequence`s. I changed it there since `debounce` should inherit the _castiness_ from it's base `AsyncSequence`